Documentation for the PostgreSQL vector store connector

lossyrob opened the pull request:
This PR adds the Postgres code samples to code-samples.md, expands on some of the dotnet docs, and adds python docs for the Postgres connector.

lossyrob commented:
@westey-m now that both the pythion and dotnet versions are released for PG vector search, it would be great to get the docs published as well. What's the process for review and publishing here? Thanks!

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -41,19 +41,43 @@ Add the Postgres Vector Store connector NuGet package to your project.
+    dataSourceBuilder.UseVector();
+    return dataSourceBuilder.Build();
+});
+kernelBuilder.AddPostgresVectorStore();
```
Only `IServiceCollection` registrations are supported.
```suggestion
kernelBuilder.Services.AddPostgresVectorStore();
```

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
-builder.Services.AddSingleton<NpgsqlDataSource>(sp => 
-{
-    NpgsqlDataSourceBuilder dataSourceBuilder = new("Host=localhost;Port=5432;Username=postgres;Password=example;Database=postgres;");
+kernelBuilder.Services.AddSingleton<NpgsqlDataSource>(sp => {
+    NpgsqlDataSourceBuilder dataSourceBuilder = new("<Connection String>");
```
```suggestion
builder.Services.AddSingleton<NpgsqlDataSource>(sp => 
{
    NpgsqlDataSourceBuilder dataSourceBuilder = new("<Connection String");
```

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
````diff
@@ -62,91 +86,420 @@ using Npgsql;
 ```csharp
 using Microsoft.SemanticKernel.Connectors.Postgres;
-using Npgsql;
+using Postgres.Client;
````
```suggestion
using Npgsql;
```

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
-
-var collection = new PostgresVectorStoreRecordCollection<string, Hotel>(dataSource, "skhotels");
+using NpgsqlDataSource dataSource = dataSourceBuilder.Build();
+var collection = new PostgresVectorStoreRecordCollection<Hotel>(dataSource, "skhotels");
```
```suggestion
var collection = new PostgresVectorStoreRecordCollection<string, Hotel>(dataSource, "skhotels");
```

lossyrob replied:
Changed this to match `int`, which I used in the Hotels class example

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+
+public static class NpgsqlDataSourceBuilderExtensions
+{
+    private static readonly TokenRequestContext s_azureDBForPostgresTokenRequestContext = new([Constants.AzureDBForPostgresScope]);
```
Can we add the constant value into the snippet of code, since `Constants` does not exist in the sample, so users won't know what the value is otherwise.

lossyrob replied:
I updated the example to match what was reviewed by the Npgsql maintainer here: https://github.com/npgsql/npgsql/issues/5992

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);
+dataSourceBuilder.UseEntraAuthentication();
+dataSourceBuilder.UseVector();
+using var datasource = dataSourceBuilder.Build();
```
```suggestion
using var dataSource = dataSourceBuilder.Build();
```

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
````diff
@@ -62,91 +86,420 @@ using Npgsql;
+Now you can use the `UseEntraAuthentication` method to set up the connection string for the Postgres connector:
+
+```csharp
+var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);
````
We should give an example of what the connection string should look like in this case, where the username and token is retrieved from elsewhere.

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
 
-More info coming soon.
+settings = PostgresSettings()
+with settings.create_connection_pool() as pool:
```
as far as I can see in the code, this is not a context manager, and it is def a async method
```suggestion
pool = await settings.create_connection_pool()
vector_store = PostgresStore(pool=pool)
```

eavanvalkenburg replied:
btw, I would be open to making this a context manager, I think that makes sense, we would have to do a PR for that though!

lossyrob replied:
This is fixed; the pool is a context manager, I was just using it incorrectly with out the async awaits.

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+from semantic_kernel.connectors.memory.Postgres import PostgresCollection
+
+collection = PostgresCollection(collection_name="skhotels", data_model_type=Hotel)
+with collection: # This will create a connection pool using PostgresSettings
```
```suggestion
async with collection: # This will create a connection pool using PostgresSettings
```

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+logger = logging.getLogger(__name__)
+
+
+async def get_entra_token_async(credential: AsyncTokenCredential) -> str:
```
in `semantic_kernel/utils/authentication/entra_id_authentication.py` we have prebuilt function for this, just pass the endpoint into that!

lossyrob replied:
That function is helpful, but there's not an async version and doesn't take in a credential as an option so always uses `DefaultAzureCredential`...passing in a credential is useful e.g. when you are operating in a service pod on AKS that has a workload identity, and you don't want to iterate through the possible credential paths because you only want a single credential mechanism.

I could contribute a PR that adds the credential argument and async version if you think that would be useful.

eavanvalkenburg replied:
that would be very useful, I'm very surprised there is no async in that one... and I like the idea of passing specific creds!

lossyrob replied:
Ok, put that on the TODO list. Looks like there's implications for the classes that use it, e.g. allowing credentials to be specificed in `AzureOpenAIConfigBase`.Do you think that should be blocking to this PR? It would be great to get this into the docs now that the postgres vectorstore feature is published. I can update the docs once the utils change gets merged and released.

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+from semantic_kernel.connectors.memory.Postgres import PostgresSettings, PostgresStore
+
+settings = PostgresSettings(connection_class=AsyncEntraConnection)
+with settings.create_connection_pool() as pool:
```
same as above

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
+from azure.identity import ManagedIdentityCredential 
+
+settings = PostgresSettings(connection_class=AsyncEntraConnection, credential=ManagedIdentityCredential())
+with settings.create_connection_pool() as pool:
```
and one more

lossyrob commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -62,91 +86,420 @@ using Npgsql;
-    "hotel_description" TEXT ,
-    "DescriptionEmbedding" VECTOR(4) ,
-    PRIMARY KEY ("HotelId")
+CREATE TABLE Hotels (
```
Actually I read the code wrong - I am doing the PRIMARY KEY correctly, but instead of specifying the constraint in the column declaration, doing it at the end of the CREATE table statement. Updated documentation to match this.

lossyrob commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -91,62 +87,423 @@ It is possible to construct a direct reference to a named collection.
 
-You can override property names to use in storage that is different to the property names on the data model.
-The property name override is done by setting the `StoragePropertyName` option via the data model property attributes or record definition.
+For data properties and vector properties (if using named vectors mode), you can provide override field names to use in storage that is different from the
```
Correct - fixed!

lossyrob commented:
@westey-m lost track of this after the approval, though it would make it in. Is there anything you need from me to get this PR included in the next docs release?

AbeOmor commented:
#sign-off

eavanvalkenburg commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
````diff
@@ -130,66 +126,422 @@ It is possible to construct a direct reference to a named collection with a cust
+or use the `PostgresSettings` to create a connection pool from environment variables.
+
+```python
+from semantic_kernel.connectors.memory.postgres import PostgresStore, PostgresSettings
````
We moved a bunch of things around so please have a look, every import should now come from semantic_kernel.connectors.postgres

lossyrob replied:
Fixed in  755d00f

lossyrob commented:
Python Documentation Updates Complete ✅

  I've updated all the Python examples in the PostgreSQL connector documentation to address the import
  reorganization feedback from @eavanvalkenburg.

  Changes Made:

  🔧 Import Path Updates:
  - Changed all imports from semantic_kernel.connectors.memory.postgres to semantic_kernel.connectors.postgres
  - Updated data model imports to use semantic_kernel.data.vector

  🏗️ API Structure Updates:
  - Fixed VectorStoreField usage to match current API (removed unsupported has_embedding parameter)
  - Updated parameter order in vector field definitions (dimensions parameter first)
  - Changed data_model_type to record_type in collection instantiation
  - Added missing typing.Annotated import

  📝 Specific Files Updated:
  - Updated all Python code examples in postgres-connector.md
  - Fixed basic usage examples (lines 349, 366)
  - Updated complete data model example (lines 384-414)
  - Fixed Entra authentication example (line 522)

  ✅ Validation:
  - All imports work correctly with the latest semantic kernel Python codebase
  - Model definitions compile and instantiate properly
  - Collection creation works as expected
  - Comprehensive testing confirms examples are functional

  The documentation now accurately reflects the current semantic kernel Python library structure where the Postgres
   connector has been moved to the simplified semantic_kernel.connectors.postgres import path.

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -77,32 +77,29 @@ dotnet add package Microsoft.SemanticKernel.Connectors.PgVector --prerelease
-// Using IServiceCollection with ASP.NET Core.
-var builder = WebApplication.CreateBuilder(args);
-builder.Services.AddPostgresVectorStore("Host=localhost;Port=5432;Username=postgres;Password=example;Database=postgres;");
+var kernelBuilder = Kernel.CreateBuilder();
```
We are de-emphasizing the use of KernelBuilder in favor of just using a service collection in SK.
Can you please follow the pattern that is in the existing docs around this, e.g. 
```csharp
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddPostgresVectorStore...
```

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -130,66 +126,422 @@ It is possible to construct a direct reference to a named collection with a cust
-
-var collection = new PostgresCollection<string, Hotel>("Host=localhost;Port=5432;Username=postgres;Password=example;Database=postgres;", "skhotels");
+using NpgsqlDataSource dataSource = dataSourceBuilder.Build();
+var collection = new PostgresVectorStoreRecordCollection<int, Hotel>(dataSource, "skhotels");
```
This class has been renamed to PostgresCollection and it has a new param, see on line 137 of the original doc.

westey-m commented on semantic-kernel/concepts/vector-store-connectors/out-of-the-box-connectors/postgres-connector.md:
```diff
@@ -130,66 +126,422 @@ It is possible to construct a direct reference to a named collection with a cust
 public class Hotel
 {
-    [VectorStoreKey]
+    [VectorStoreRecordKey(StorageName = "hotel_id")]
```
Also see the original doc, the names of these attributes have changed.
